src/parser/core/DotSnapshots.ts

```diff
@@ -0,0 +1,279 @@
+    const previousSnapshots = previousUptime ? previousUptime.snapshots : null;
+    const prevPower =
+      previousSnapshots === null
+        ? 0
+        : previousSnapshots.reduce((acc, ss) => acc * (1 + ss.boostStrength), 1);
+
+    this.handleApplication(
+      event,
+      snapshots,
+      previousSnapshots,
+      power,
+      prevPower,
+      remainingOnPrev,
+      clipped,
+    );
+
+    uptimes.push({
+      start: event.timestamp,
+      expectedEnd,
+      snapshots,
+      previousSnapshots,
+    });
+  }
+
+  _finishDot(event: RemoveDebuffEvent | RefreshDebuffEvent): DotUptime | undefined {
+    const uptimes = this.getUptimesForTarget(event);
+    if (uptimes.length > 0) {
+      const latestUptime = uptimes[uptimes.length - 1];
+      latestUptime.end = event.timestamp;
+      return latestUptime;
+    } else {
+      console.warn(event.type + ' on target when we have no record of an applydebuff', event);
+      return undefined;
+    }
+  }
+
+  _getActiveSnapshots(timestamp: number): SnapshotSpec[] {
+    return this.applicableSnapshots.filter((as) => as.isPresent(this.selectedCombatant, timestamp));
+  }
+
+  get snapshotUptimes(): UptimeBarSpec[] {
+    return this.applicableSnapshots.map((as) => this._getSnapshotUptimesForBuff(as));
+  }
+
+  /** Gets the specific uptime periods where at least one active DoT has the given snapshot */
+  getSnapshotCombinedUptimes(snapshotName: string): ClosedTimePeriod[] {
+    return mergeTimePeriods(
+      Object.values(this.snapshotsByTarget)
+        .flatMap((uptimes) => uptimes)
+        .filter((uptime) => uptime.snapshots.find((ss) => ss.name === snapshotName) !== undefined),
+      this.owner.currentTimestamp,
+    );
+  }
+
+  /** Gets the duration in ms where at least one active DoT has the given snapshot */
+  getSnapshotCombinedUptimeDuration(snapshotName: string): number {
+    return this.getSnapshotCombinedUptimes(snapshotName).reduce(
+      (acc, up) => acc + up.end - up.start,
+      0,
+    );
+  }
+
+  /** Gets the percent of the DoT's total uptime that a snapshotted DoT was active on at least one target */
+  getPercentUptimeWithSnapshot(snapshotName: string): number {
+    const totalUptime = this.getTotalDotUptime();
+    return totalUptime === 0
+      ? 0
+      : this.getSnapshotCombinedUptimeDuration(snapshotName) / totalUptime;
+  }
+
+  /** Gets the time remaining on the DoT active on the given event's target at the time of the event (zero if DoT isn't active) */
+  getTimeRemaining(event: TargettedEvent<any>): number {
+    const uptimes = this.getUptimesForTarget(event);
+    if (uptimes.length > 0) {
+      const latestUptime = uptimes[uptimes.length - 1];
+      return Math.max(0, latestUptime.expectedEnd - event.timestamp);
+    } else {
+      return 0;
+    }
+  }
+
+  /** Builds an uptime bar for the given shapshot spec */
+  _getSnapshotUptimesForBuff(spec: SnapshotSpec): UptimeBarSpec {
```

Comment:
```suggestion
  private _getSnapshotUptimesForBuff(spec: SnapshotSpec): UptimeBarSpec {
```

(or protected if this is supposed to be called by implementors)